pkg/gofr/datasource/surrealdb/surrealdb.go

```
@@ -0,0 +1,543 @@
+	if m, ok := metrics.(Metrics); ok {
+		c.metrics = m
+	}
+}
+
+// UseTracer sets a custom tracer for the Client.
+func (c *Client) UseTracer(tracer interface{}) {
+	if t, ok := tracer.(trace.Tracer); ok {
+		c.tracer = t
+	}
+}
+
+// Connect establishes a connection to the SurrealDB server using the client's configuration.
+var (
+	errNoDatabaseInstance       = errors.New("failed to connect to SurrealDB: no valid database instance")
+	errInvalidCredentialsConfig = errors.New("both username and password must be provided")
+	errEmbeddedDBNotEnabled     = errors.New("embedded database not enabled")
+	errInvalidConnectionURL     = errors.New("invalid connection URL")
+)
+
+// NewDB creates a new SurrealDB client.
+func NewDB(connectionURL string) (con connection.Connection, err error) {
+	u, err := url.ParseRequestURI(connectionURL)
+	if err != nil {
+		return nil, err
+	}
+
+	scheme := u.Scheme
+
+	newParams := connection.NewConnectionParams{
+		Marshaler:   models.CborMarshaler{},
+		Unmarshaler: models.CborUnmarshaler{},
+		BaseURL:     fmt.Sprintf("%s://%s", u.Scheme, u.Host),
+		Logger:      logger.New(slog.NewTextHandler(os.Stdout, nil)),
+	}
+
+	if scheme == "http" || scheme == https {
+		con = connection.NewHTTPConnection(newParams)
+	} else if scheme == ws || scheme == "wss" {
+		con = connection.NewWebSocketConnection(newParams)
+	} else if scheme == "memory" || scheme == "mem" || scheme == "surrealkv" {
+		return nil, fmt.Errorf("%w", errEmbeddedDBNotEnabled)
+	} else {
+		return nil, fmt.Errorf("%w", errInvalidConnectionURL)
+	}
+
+	err = con.Connect()
+	if err != nil {
+		return nil, err
+	}
+
+	return con, nil
+}
+
+// Connect establishes a connection to the SurrealDB database.
+func (c *Client) Connect() {
+	endpoint := c.buildEndpoint()
+	err := c.connectToDatabase(endpoint)
+
+	if err != nil {
+		return
+	}
+
+	err = c.setupNamespaceAndDatabase()
+	if err != nil {
+		return
+	}
+
+	err = c.authenticate()
+	if err != nil {
+		return
+	}
```

Errors are not returned because of an old interface design I regret. There is nothing you can do in this PR.

**But** you should log these errors, it cannot be silent

---

I am already logging these errors using c.logError() in my implementation. in setupNamespaceAndDatabase()
```
func (c *Client) setupNamespaceAndDatabase() error {
	err := c.db.Use(c.config.Namespace, c.config.Database)
	if err != nil {
		c.logError("unable to set the namespace and database for SurrealDB", err)
		return err
	}

	return nil
}
```